pkg/iprange/parse.go

```diff
@@ -0,0 +1,144 @@
+	if len(parts) == 0 {
+		return nil, nil
+	}
+
+	var ranges []Range
+	for _, v := range parts {
+		r, err := ParseRange(v)
+		if err != nil {
+			return nil, err
+		}
+
+		if r != nil {
+			ranges = append(ranges, r)
+		}
+	}
+	return ranges, nil
+}
+
+var (
+	reRange      = regexp.MustCompile("^[0-9a-f.:-]+$")            // addr | addr-addr
+	reCIDR       = regexp.MustCompile("^[0-9a-f.:-]+/[0-9]{1,3}$") // addr/prefix_length
+	reSubnetMask = regexp.MustCompile("^[0-9.-]+/[0-9.]{7,}$")     // v4_addr/mask
+)
+
+// ParseRange parses s as an IP Range, returning the result and an error if any.
+// The string s can be in IPv4 address ("192.0.2.1"), IPv4 range ("192.0.2.0-192.0.2.10")
+// IPv4 CIDR ("192.0.2.0/24"), IPv4 subnet mask ("192.0.2.0/255.255.255.0"),
+// IPv6 address ("2001:db8::1"), IPv6 range ("2001:db8::-2001:db8::10"),
+// or IPv6 CIDR ("2001:db8::/64") form.
+// IPv4 CIDR, IPv4 subnet mask and IPv6 CIDR ranges don't include host and network IPs.
+func ParseRange(s string) (Range, error) {
+	s = strings.ToLower(s)
+	if s == "" {
+		return nil, nil
+	}
+
+	var r Range
+	switch {
+	case reRange.MatchString(s):
+		r = parseRange(s)
+	case reCIDR.MatchString(s):
+		r = parseCIDR(s)
+	case reSubnetMask.MatchString(s):
+		r = parseSubnetMask(s)
+	}
+
+	if r == nil {
+		return nil, fmt.Errorf("ip range (%s) invalid syntax", s)
+	}
+	return r, nil
+}
+
+func parseRange(s string) Range {
+	var start, end net.IP
+	if idx := strings.IndexByte(s, '-'); idx != -1 {
+		start, end = net.ParseIP(s[:idx]), net.ParseIP(s[idx+1:])
+	} else {
+		start, end = net.ParseIP(s), net.ParseIP(s)
+	}
+
+	switch {
+	case isV4RangeValid(start, end):
+		return v4Range{start: start, end: end}
+	case isV6RangeValid(start, end):
+		return v6Range{start: start, end: end}
+	default:
+		return nil
+	}
+}
+
+func parseCIDR(s string) Range {
+	ip, network, err := net.ParseCIDR(s)
+	if err != nil {
+		return nil
+	}
+
+	start, end := cidr.AddressRange(network)
+	prefixLen, _ := network.Mask.Size()
+
+	if isV4IP(ip) && prefixLen < 31 || isV6IP(ip) && prefixLen < 127 {
+		start = cidr.Inc(start)
+		end = cidr.Dec(end)
+	}
+
+	return parseRange(fmt.Sprintf("%s-%s", start, end))
+}
+
+func parseSubnetMask(s string) Range {
+	idx := strings.LastIndexByte(s, '/')
+	if idx == -1 {
+		return nil
+	}
+
+	address, mask := s[:idx], s[idx+1:]
+
+	ip := net.ParseIP(mask).To4()
+	if ip == nil {
+		return nil
+	}
+
+	prefixLen, bits := net.IPv4Mask(ip[0], ip[1], ip[2], ip[3]).Size()
+	if prefixLen+bits == 0 {
+		return nil
+	}
+
+	return parseCIDR(fmt.Sprintf("%s/%s", address, strconv.Itoa(prefixLen)))
+}
+
+func isV4RangeValid(start, end net.IP) bool {
+	return isV4IP(start) && isV4IP(end) && bytes.Compare(end, start) >= 0
+}
+
+func isV6RangeValid(start, end net.IP) bool {
+	return isV6IP(start) && isV6IP(end) && bytes.Compare(end, start) >= 0
+}
+
+func isV4IP(ip net.IP) bool {
+	return ip.To4() != nil
+}
+
+func isV6IP(ip net.IP) bool {
+	return !isV4IP(ip) && ip.To16() != nil
```

Comment:
A minor thing here, you tested the return from `isV4IP`  using a different pattern in the whole code until now. This comparison is not wrong, but it is not following the pattern.

Reply:
what pattern?

Reply:
You wrote for the majority of the code `function(value) != nil` instead `!function(value)`.

Reply:
it is not pattern, i could write it as `ip.To4() == nil && ip.To16() != nil` but since i have `isV4IP` i used it

std lib source code
```go
// To4 converts the IPv4 address ip to a 4-byte representation.
// If ip is not an IPv4 address, To4 returns nil.
func (ip IP) To4() IP {
	...
}

// To16 converts the IP address ip to a 16-byte representation.
// If ip is not an IP address (it is the wrong length), To16 returns nil.
func (ip IP) To16() IP {
	...
}
```

So `To16()` returns **not nil** for both v4 and v6, `To4()` returns **not nil** only for v4.